chore(RHINENG-2211): React 18 & Patternfly 5 Upgrades

LiorKGOW opened the pull request:
# Description

This PR upgrades this application to React 18 & Patternfly 5 (along with some necessary packages)

Associated Jira tickets: 
[#RHINENG-2211](https://issues.redhat.com/browse/RHINENG-2211)
[#RHINENG-5939](https://issues.redhat.com/browse/RHINENG-5939)

As mentioned in [this slack thread](https://redhat-internal.slack.com/archives/C02T90PLJ5V/p1705484895906979), we agreed on not upgrading `redhat-cloud-services/frontend-components-config`, 
therefore I did not change it in this PR

# How to test the PR

Have a look that the application works as expected

# Checklist:

- [x] The commit message has the Jira ticket linked
- [x] PR has a short description
- [x] Upgrade to React 18
- [x] Upgrade to Patternfly 5
- [X] Fix failing tests
- [x] Fix commits

LiorKGOW committed:
(chore): Update packages before React 18 upgrade

LiorKGOW committed:
(chore): Changes from running pf-codemod

run this command
`npx @patternfly/pf-codemods@latest . --fix`
it was specified here:
https://www.patternfly.org/get-started/upgrade/

LiorKGOW committed:
(chore): Changes from running pf-codemod

Ran the command: `npx @patternfly/pf-codemods@latest ./ --fix`

LiorKGOW committed:
(fix): Fix errors from the pf-codemod

jmbuscemi commented:
I would just like to request that Teamnado's QE team gets a chance to run our Activation Key regression tests once this makes it's way to Stage and before it goes to Prod so that we can assure nothing there is broken.

You can ping myself or Luke in our Slack channel and we'll get the QE team to run the tests.

LiorKGOW committed:
(feat): Upgrade to React 18

LiorKGOW committed:
(feat): Upgrade to Patternfly 5

LiorKGOW committed:
(fix): Fix failing tests & update snapshots

LiorKGOW committed:
(chore): Update Client Rendering APIs for React 18

Followed instructions specified on react.dev:
https://react.dev/blog/2022/03/08/react-18-upgrade-guide

LiorKGOW committed:
(chore): Upgrade react testing library packages

LiorKGOW commented:
> I would just like to request that Teamnado's QE team gets a chance to run our Activation Key regression tests once this makes it's way to Stage and before it goes to Prod so that we can assure nothing there is broken.
> 
> You can ping myself or Luke in our Slack channel and we'll get the QE team to run the tests.

I will ping you once this PR gets merged @jmbuscemi 👍🏼

LiorKGOW committed:
(chore): remove react-hooks and changing imports

https://github.com/testing-library/react-hooks-testing-library?tab=readme-ov-file#a-note-about-react-18-support

LiorKGOW committed:
(chore): Fix failing test

LiorKGOW committed:
(chore): update package minor version

LiorKGOW committed:
(fix): Fix TextInput's onChange

LiorKGOW committed:
(feat): Refactor removed cypress test to RTL

LiorKGOW committed:
(chore): fix failing test & update snapshot

LiorKGOW committed:
(chore): fix failing tests

+ removing comments which could be amended to a previous commit

https://github.com/stalniy/bdd-lazy-var/issues/97#issuecomment-1125929739

LiorKGOW committed:
(chore): Upgrading react-query to v4

https://tanstack.com/query/v4/docs/framework/react/guides/migrating-to-react-query-4

+ 5 errors
added some comments

LiorKGOW committed:
(fix): Fix breaking changes of renderHook

LiorKGOW committed:
(feat): Refactor test

- Fix wrong getUser mock
- Remove beforeEach

LiorKGOW committed:
feat(RHINENG-2211): Upgrade to React 18 & PF 5

Link to tickets:
https://issues.redhat.com/browse/RHINENG-2211
https://issues.redhat.com/browse/RHINENG-5939

LiorKGOW committed:
(chore): reinstall removed dependencies + run lint

Also remove unnecessary comments and add final fixes

LiorKGOW commented:
This PR is ready for review ✌🏼

bastilian commented on src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:
```diff
@@ -3,54 +3,70 @@
         Contact your organization administrator(s) for more information or visit 
         <a
-          href="./settings/my-user-access"
+          href="./iam/my-user-access"
```
Is this change intended?

bastilian replied:
hm. this is a snapshot. I'm wondering where this comes from and if this is also in the actually rendered UI. And if it is actually correct.

LiorKGOW replied:
let me have a look

LiorKGOW replied:
@bastilian This change came from this commit:
https://github.com/RedHatInsights/sed-frontend/pull/272/commits/1d4fe7cdbae4c010fa53ffc1f02cf6a58e6ded7a 
But I did not make any changes to `Authentication.js` here so I guess it came from a different commit and I just updated the snapshot here

I checked previous commits, and I only made relevant changes in those commits:
1. This is the only commit where changes were made to `Authentication.js` & `Authentication.test.js` files, but those changes came after the snapshot update you mentioned, so they couldn't have came up from this commit: https://github.com/RedHatInsights/sed-frontend/pull/272/commits/53a5bf9dbfe01af4f72327af0ccd5ddd6c455d4f#diff-a194d08568455b4d77eb5df13c98260410d3ed5375a39aee43331d657dd85489 
2. Here was the first snapshot update to the snapshot file you mentioned, but I did not make any previous changes to `Authentication.js` & `Authentication.test.js` files, so it must have been caused from a previous package update, but I am not sure which one caused this change.
https://github.com/RedHatInsights/sed-frontend/pull/272/commits/f01bac6dd852a4b70b68d4b23200ea39f1349426#diff-f6381abb627b767d59d1844907725850526ea230abf57154c7492a365c0e5a20

Maybe this change came from the `config/setupTests.js` change in the same commit ? or a previous package update ?

LiorKGOW replied:
I found it in the UI

This is the page it is found on:
![image](https://github.com/RedHatInsights/sed-frontend/assets/93318917/76951840-6634-4e18-959d-0e18829c963d)

And this is the link it sends you to when you click [My User Access](https://stage.foo.redhat.com:1337/iam/my-user-access):
![image](https://github.com/RedHatInsights/sed-frontend/assets/93318917/08555441-2025-4e3e-9337-75fdf929c996)

It seems to me that this is the functionality we want when we click this link @bastilian

jmbuscemi replied:
That first screenshot is what we need to show in Activation Keys when a user doesn't have access.

If you need help with any of the Activation Key pieces let us know and we can take a look.

LiorKGOW replied:
@jmbuscemi Does the link work as expected ? 
We were just wondering if the code works as it should (because the link in the snapshot changed)

jmbuscemi replied:
@jlgre when you get some time can you look into Activation Keys w.r.t. this PR?

lukegrn replied:
I believe that is the correct link  - I don't even think that's a component that we control.

settings/my-user-access breaks anyway as far as I can see, so the iam one seems right

mkholjuraev commented on .cache_iifckc:
I believe we can remove this cache file generated by pf-codemods as it will not be used anymore

LiorKGOW replied:
I removed it 👍🏼 
Thanks @mkholjuraev

LiorKGOW committed:
(chore): Resolve comments from reviewers

Remove cache file generated by pf-codemods

LiorKGOW committed:
(chore): Fix lint errors

Ignored warnings because they are all about imports

lukegrn commented:
Haven't reviewed the code yet, but upon running locally I see an error in the activation keys create wizard. When the 4th step is entered and something has been selected for SLA on the 3rd step, the modal crashes upon step advance.

gkarat commented:
@LiorKGOW, I tried to reproduce the bug that @jlgre found and I see that there is a non-string value we set for `sla` through `setSla` function: https://github.com/LiorKGOW/sed-frontend/blob/react-18-upgrade/src/Components/Pages/SetSystemPurposePage.js#L73. I checked the new interface for `onChange` for FormSelect and there is indeed the first argument is of event type. So this would need to be rewritten to smth like:
 ```
            <FormSelect onChange={(event, value) => setSla(value)} value={sla} id="activation-key-sla">
```

gkarat commented:
The warnings in the console can be helpful and indicate if there are some props of a wrong type passed to components/functions.

LiorKGOW committed:
(fix): Change onChange arguments of FormSelect

In PF5 FormSelect's onChange the first value that the function receives is an event object and not value

LiorKGOW committed:
(fix): fix console errors

LiorKGOW commented:
Thank you for having a look @gkarat @jlgre 
I have addressed the issue you mentioned @jlgre and added the changes you suggested @gkarat in this commit: [(fix): Change onChange arguments of FormSelect](https://github.com/RedHatInsights/sed-frontend/pull/272/commits/472f1d18883adc4822adbe4aae864950973f66ec)

Also, I fixed two console errors in the last commit

The PR is ready for another review 😄

LiorKGOW commented:
BTW, I thought we can also do this in the following way:
Define a more generic function that gets the necessary setter function and returns a fitting onChange function and provide it with the necessary setter function for each FormSelect
```
  const createOnChangeFunc = (setterFunc) => {
    return (_event, value) => setterFunc(value);
  };
  .
  .
  .
  <FormSelect
              onChange={createOnChangeFunc(setRole)}
              value={role}
              id="activation-key-role"
            >
```
What do you think @gkarat ?

mkholjuraev commented on src/Components/ActivationKey/SystemPurposeCard.js:
```diff
@@ -41,7 +40,21 @@ const SystemPurposeCard = (props) => {
+            </>
+          ),
+          hasNoOffset: false,
+          className: undefined,
```
Any reason to have className: undefined?

LiorKGOW replied:
No haha 
My guess is these classNames came from the 'pf-codemod' changes

I changed the classNames to be more meaningful

mkholjuraev commented on src/Components/ActivationKey/WorkloadCard.js:
```diff
@@ -32,7 +31,17 @@ const WorkloadCard = (props) => {
+            </>
+          ),
+          hasNoOffset: false,
+          className: undefined,
```
here as well

mkholjuraev reviewed (commented):
there is this one that needs a fix before we can merge the PR. When I try to add a new repo to my activation keys and try to filter by name from the dropdown, the modal gets closed unexpectedly or the page gets broken. Ref: https://redhat-internal.slack.com/archives/C02T90PLJ5V/p1708335619534349?thread_ts=1708267864.925449&cid=C02T90PLJ5V

LiorKGOW committed:
(fix): Change onChange arguments of SearchInput

In PF5 the first argument of SearchInput is the event and the second is the changed value

LiorKGOW committed:
(chore): Replace undefined classNames to more meaningful names

LiorKGOW commented:
Thank you for the review @mkholjuraev ! 

I have fixed the issue you mentioned when trying to filter the repositories by name, it was just a small change in the onChange function. In PF5 SearchInput, when the onChange function is called, the first argument is the event and the second argument is the changed value. 
It should work now

Also, I changed the classNames to be more meaningful
My guess is these classNames came from the 'pf-codemod' changes

The PR is ready for another review :)

mkholjuraev reviewed (approved):
LGTM! I had another quick look and did not encounter issues. @LiorKGOW Thank you! 

Tests need a fix though

LiorKGOW commented:
/retest

LiorKGOW committed:
(chore): Remove Cypress testing from the pipeline

LiorKGOW commented:
I removed Cypress from the pipeline because there are no Cypress tests anymore (I converted the last one to RTL)
The test is passing now @mkholjuraev 🥳

LiorKGOW commented:
@jmbuscemi The PR is merged
You can now find this on stage
